packages/back-end/src/services/features.ts

```diff
@@ -1047,3 +1151,255 @@ export const updateInterfaceEnvSettingsFromApiEnvSettings = (
+      const {
+        removeRule,
+        newPrerequisites,
+      } = getInlinePrerequisitesReductionInfo(
+        rule.prerequisites || [],
+        features,
+        environment,
+        prereqStateCache
+      );
+      if (!removeRule) {
+        rule.prerequisites = newPrerequisites;
+        newFeatureRules.push(rule);
+      }
+    }
+    newFeatures[i].environmentSettings[environment].rules = newFeatureRules;
+  }
+
+  return newFeatures;
+};
+
+export const reduceExperimentsWithPrerequisites = (
+  visualExperiments: VisualExperiment[],
+  features: FeatureInterface[],
+  environment: string,
+  prereqStateCache: Record<string, Record<string, PrerequisiteStateResult>> = {}
+): VisualExperiment[] => {
+  if (!prereqStateCache[environment]) {
+    prereqStateCache[environment] = {};
+  }
+
+  const newVisualExperiments: VisualExperiment[] = [];
+  for (const visualExperiment of visualExperiments) {
+    const phaseIndex = visualExperiment.experiment.phases.length - 1;
+    const phase: ExperimentPhase | null =
+      visualExperiment.experiment.phases?.[phaseIndex] ?? null;
+    if (!phase) continue;
+    const newVisualExperiment = cloneDeep(visualExperiment);
+
+    const {
+      removeRule,
+      newPrerequisites,
+    } = getInlinePrerequisitesReductionInfo(
+      phase.prerequisites || [],
+      features,
+      environment,
+      prereqStateCache
+    );
+    if (!removeRule) {
+      newVisualExperiment.experiment.phases[
+        phaseIndex
+      ].prerequisites = newPrerequisites;
+      newVisualExperiments.push(newVisualExperiment);
+    }
+  }
+  return newVisualExperiments;
+};
+
+export const getInlinePrerequisitesReductionInfo = (
+  prerequisites: FeaturePrerequisite[],
+  features: FeatureInterface[],
+  environment: string,
+  prereqStateCache: Record<string, Record<string, PrerequisiteStateResult>> = {}
+): {
+  removeRule: boolean;
+  newPrerequisites: FeaturePrerequisite[];
+} => {
+  if (!prereqStateCache[environment]) {
+    prereqStateCache[environment] = {};
+  }
+
+  let removeRule = false;
+  const newPrerequisites: FeaturePrerequisite[] = [];
+
+  for (const pc of prerequisites) {
+    const prereqFeature = features.find((f) => f.id === pc.id);
+    let state: PrerequisiteStateResult = {
+      state: "deterministic",
+      value: null,
+    };
+    if (prereqStateCache[environment][pc.id]) {
+      state = prereqStateCache[environment][pc.id];
+    } else {
+      if (prereqFeature) {
+        state = evaluatePrerequisiteState(
+          prereqFeature,
+          features,
+          environment,
+          undefined,
+          true
+        );
+      }
+      prereqStateCache[environment][pc.id] = state;
+    }
+
+    switch (state.state) {
+      case "conditional":
+        // keep the rule and prerequisite
+        break;
+      case "cyclic":
+        // remove the rule
+        removeRule = true;
+        continue;
+      case "deterministic":
```

Comment:
The logic in this case seems like it should be its own function:

```ts
// Given a condition and value, determine if it always passes, always fails, or is conditional
evalPrereqDeterministicValue(condition, value): "pass" | "fail" | "conditional";
```

Would be easier to unit test and make the code easier to read.

Reply:
implemented using evalCondition(). We can handle all cases now, not just simple.